lib/Backend/Vault/Kv.hs

```diff
@@ -0,0 +1,234 @@
+
+    case x of
+      WriteSecret entry -> do
+        let cofferSpecials = CofferSpecials
+                             { _masterKey =
+                               entry ^. E.masterField <&> E.getFieldKey
+                             , _globalDateModified = entry ^. E.dateModified
+                             , _fields =
+                               entry ^. E.fields & traverse %~
+                                 (\f ->
+                                 FieldMetadata
+                                 { _dateModified = f ^. E.dateModified
+                                 , _private = f ^. E.private
+                                 })
+                               & HS.mapKeys E.getFieldKey
+                             , _tags = map E.getEntryTag $ entry ^. E.tags
+                             }
+        let secret = I.PostSecret
+              { I._cas = Nothing
+              , I._pdata =
+                    HS.insert "#$coffer" (TL.toStrict . TL.decodeUtf8 $ A.encode cofferSpecials)
+                  . HS.map (^. E.value)
+                  . HS.mapKeys E.getFieldKey
+                  $ entry ^. E.fields
+              }
+
+        response <- embedCatch (entry ^. E.path) (postSecret env (entry ^. E.path) secret)
+
+        case response ^. I.ddata . at ("version" :: T.Text) of
+          Just (A.Number i) -> maybe (throw MarshallingFailed) pure (S.toBoundedInteger i)
+          _ -> throw MarshallingFailed
+      ReadSecret path version ->
+        runMaybe $ embedCatchMaybe path (readSecret env path version)
+        >>= \(I.KvResponse _ _ _ _ (I.ReadSecret _data _ _ _ _ _) _ _ _) -> do
+          cofferSpecials :: CofferSpecials <-
+            maybeThrow (_data ^.at "#$coffer" >>= A.decodeStrict' . T.encodeUtf8)
+          let secrets = HS.toList $ foldr HS.delete _data ["#$coffer"]
+          let keyToField key = do
+               _modTime <- cofferSpecials ^. fields.at key <&> (^. dateModified)
+               _private <- cofferSpecials ^. fields.at key <&> (^. private)
+               _value <- _data ^.at key
+               _key <- E.newFieldKey key
+
+               Just (_key
+                    , E.newField _modTime _value
+                      & E.private .~ _private
+                    )
+
+          fields <- maybeThrow $
+            secrets & traverse %~ (keyToField . fst) & sequence <&> HS.fromList
+          _tags <- maybeThrow $ cofferSpecials ^. tags & mapM E.newEntryTag
+
+          pure $ E.newEntry path (cofferSpecials ^. globalDateModified)
+            & E.masterField .~ (cofferSpecials ^. masterKey >>= E.newFieldKey)
+            & E.fields .~ fields
+            & E.tags .~ _tags
+      ListSecrets path ->
+        runMaybe (embedCatchMaybe path (listSecrets env path) <&> (^. I.ddata) <&> \(I.ListSecrets list) -> list)
+      DeleteSecret path ->
+        embedCatch path (deleteSecret env path) <&> const ()
+  where postSecret env = (I.routes env ^. I.postSecret) mount token
```

Comment:
1) We have styleguide; 
2) that `where` is a major offence to the reader.

Reply:
Done